File: j2kt/jre/java/native/java/lang/Throwable.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package java.lang;

import com.google.j2kt.annotations.HiddenFromObjC;
import java.io.PrintStream;
import java.io.PrintWriter;
import javaemul.internal.annotations.KtNative;
import javaemul.internal.annotations.KtProperty;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@KtNative(
    name = "kotlin.Throwable",
    bridgeName = "javaemul.lang.ThrowableJvm",
    companionName = "java.lang.Throwable")
@NullMarked
public class Throwable {

  public Throwable() {}

  public Throwable(@Nullable String detailMessage) {}

  public Throwable(@Nullable String detailMessage, @Nullable Throwable cause) {}

  public Throwable(@Nullable Throwable cause) {}

  protected Throwable(
      @Nullable String detailMessage,
      @Nullable Throwable cause,
      boolean enableSuppression,
      boolean writableStackTrace) {}

  public native Throwable fillInStackTrace();

  @KtProperty
  public native @Nullable String getMessage();

  @KtProperty
  public native @Nullable String getLocalizedMessage();

  // Treat as a property to avoid conflict with `fun getStacktrace: Array<String>` on Kotlin/Native.
  @KtProperty
  public native StackTraceElement[] getStackTrace();

  public native void setStackTrace(StackTraceElement[] trace);

  /**
   * Writes a printable representation of this {@code Throwable}'s stack trace to the {@code
   * System.err} stream.
   */
  public native void printStackTrace();

  public native void printStackTrace(PrintStream err);

  public native void printStackTrace(PrintWriter err);

  @Override
  public native String toString();

  public native Throwable initCause(@Nullable Throwable throwable);

  @KtProperty
  public native @Nullable Throwable getCause();

  public final native void addSuppressed(Throwable throwable);

  @KtProperty
  public final native Throwable[] getSuppressed();

  // Method existing in J2CL-JRE and referenced by user code. This method is not available in
  // j2kt-native and j2kt-jvm.
  @HiddenFromObjC
  public static native Throwable of(@Nullable Object e);
}

File: j2kt/jre/java/native/java/lang/Throwable.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package java.lang

import java.io.PrintStream
import java.io.PrintWriter
import javaemul.lang.CauseHolder
import javaemul.lang.InitCauseCapable
import kotlin.String as KotlinString
import kotlin.Throwable as KotlinThrowable

open class Throwable(message: KotlinString? = null) : KotlinThrowable(message), InitCauseCapable {
  override val causeHolder = CauseHolder()
  override val cause
    get() = causeHolder.cause

  constructor(cause: KotlinThrowable?) : this() {
    initCause(cause)
  }

  constructor(message: KotlinString?, cause: KotlinThrowable?) : this(message) {
    initCause(cause)
  }

  constructor(
    message: KotlinString?,
    cause: KotlinThrowable?,
    enableSuppression: Boolean,
    writableStackTrace: Boolean,
  ) : this(message) {
    initCause(cause)
  }

  override fun initCause(cause: KotlinThrowable?): KotlinThrowable =
    super<InitCauseCapable>.initCause(cause)

  open fun printStackTrace(stream: PrintStream) {
    stream.print(stackTraceToString())
  }

  open fun printStackTrace(writer: PrintWriter) {
    writer.write(stackTraceToString())
  }

  open fun fillInStackTrace(): KotlinThrowable = this

  // TODO(b/276846862): We may want to wrap native stack trace elements in StackTraceElement
  //   and emulating the "original" StackTraceElement as far as possible.
  open val stackTrace: Array<StackTraceElement>
    get() = emptyArray<StackTraceElement>()

  open fun setStackTrace(trace: Array<StackTraceElement>) {}

  open val localizedMessage: KotlinString?
    get() = message
}
